File: 1620.coordinate-with-maximum-network-quality.py
#
# @lc app=leetcode id=1620 lang=python3
#
# [1620] Coordinate With Maximum Network Quality
#

# @lc code=start
import math
from typing import List


class Solution:
    # 1 <= towers.length <= 50
    # 0 <= xi, yi, qi <= 50
    # So just brute force
    
    def euclidean(self, x1, y1, x2, y2):
        return math.sqrt((x1-x2)*(x1-x2) + (y1-y2)*(y1-y2))
    
    def bestCoordinate(self, towers: List[List[int]], radius: int) -> List[int]:
        quality = [[0 for j in range(51)] for i in range(51)]
        for x, y, q in towers:
            for i in range(51):
                for j in range(51):
                    d = self.euclidean(i, j, x, y)
                    if d <= radius:
                        quality[i][j] += math.floor(q / (1+d))
        
        x, y, qMax = 0, 0, 0
        for i in range(51):
            for j in range(51):
                if quality[i][j] > qMax:
                    qMax = quality[i][j]
                    x, y = i, j
        return [x, y]
# @lc code=end

File: 1620.coordinate-with-maximum-network-quality.cs
/*
 * @lc app=leetcode id=1620 lang=csharp
 *
 * [1620] Coordinate With Maximum Network Quality
 */

// @lc code=start
public class Solution 
{
    public double euclidean(int x1, int y1, int x2, int y2)
    {
        return Math.Sqrt((x1 - x2)*(x1 - x2) + (y1 - y2)*(y1 - y2));
    }
    public int[] BestCoordinate(int[][] towers, int radius) 
    {
        int[,] quality = new int[51,51];
        for (int i = 0; i < 51; i++)
        {
            for (int j = 0; j < 51; j++)
            {
                quality[i,j] = 0;
                foreach (int[] tower in towers)
                {
                    double d = euclidean(i, j, tower[0], tower[1]);
                    if (d <= radius)
                    {
                        quality[i,j] += (int)Math.Floor(tower[2] / (1 + d));
                    }
                }
            }
        }
        int x = 0;
        int y = 0;
        int qMax = 0;
        for (int i = 0; i < 51; i++)
        {
            for (int j = 0; j < 51; j++)
            {
                if (quality[i,j] > qMax)
                {
                    qMax = quality[i,j];
                    x = i;
                    y = j;
                }
            }
        }
        return new int[2]{x,y};
    }
}
// @lc code=end

